pkg/chainreader/contract_reader.go

```diff
@@ -0,0 +1,111 @@
+package chainreader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+// ContractReaderByIDs wraps types.ContractReader to allow the caller to set custom contractIDs with Bind.
+type ContractReaderByIDs interface {
+	// Bind accepts a map of bound contracts where map keys are custom contractIDs to be used for calling the interface.
+	Bind(ctx context.Context, bindings map[string]types.BoundContract) error
+	Unbind(ctx context.Context, bindings map[string]types.BoundContract) error
+	GetLatestValue(ctx context.Context, contractID, readName string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error
+	BatchGetLatestValues(ctx context.Context, request BatchGetLatestValuesRequestByCustomID) (BatchGetLatestValuesResultByCustomID, error)
+	QueryKey(ctx context.Context, contractID string, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]types.Sequence, error)
+}
+
+// WrapContractReaderByIDs returns types.ContractReader behind ContractReaderByIDs interface.
+func WrapContractReaderByIDs(contractReader types.ContractReader) ContractReaderByIDs {
+	return &contractReaderByIDs{
+		cr: contractReader,
+	}
+}
+
+type contractReaderByIDs struct {
+	bindings sync.Map
+	cr       types.ContractReader
+}
+
+type BatchGetLatestValuesRequestByCustomID map[string]types.ContractBatch
+type BatchGetLatestValuesResultByCustomID map[string]types.ContractBatchResults
+
+func (crByIds *contractReaderByIDs) Bind(ctx context.Context, bindings map[string]types.BoundContract) error {
+	var toBind []types.BoundContract
+	for customContractID, boundContract := range bindings {
+		crByIds.bindings.Store(customContractID, boundContract)
+		toBind = append(toBind, boundContract)
+	}
+	return crByIds.cr.Bind(ctx, toBind)
+}
+
+func (crByIds *contractReaderByIDs) Unbind(ctx context.Context, bindings map[string]types.BoundContract) error {
+	var toUnbind []types.BoundContract
+	for customContractID, boundContract := range bindings {
+		crByIds.bindings.Delete(customContractID)
+		toUnbind = append(toUnbind, boundContract)
+	}
+	return crByIds.cr.Unbind(ctx, toUnbind)
```

Comment:
If there's an error, could `crByIds` be in a bad state?

Reply:
I think with the current EVM impl. even in the worst scenario (LP having a dangling filter tracking a contract for no reason) this would be fine. I changed it though to future proof a bit more, we should make Unbind completely transactional so that the behaviour is easier to predict and handle when we get to multiple chain impl.